### ISSUE TITLE
New simulator example for Mount, Camera and Focuser. Update for cu_alpacadevice for 2d images

### ISSUE THREAD

**Issue by han-k59** [pchev/alpaca_server#1]
Thanks for the code!

Attached my simple simulator example and updated server files. The simulator will create an artificial image which can be read by CCDCiel. For the focuser two mods are required in CCDciel. See bug tracker CCDCiel.

I will now work on my "Simulator for ASCOM" using this code.

Han

[example simulator.zip](https://github.com/pchev/alpaca_server/files/7262361/example.simulator.zip)
[alpaca_server.zip](https://github.com/pchev/alpaca_server/files/7259680/alpaca_server.zip)

later, added missing line MyFocuser.SetConnected(False); in the example:

**Comment by pchev** [comment]
Thank you!

I add and commit the code from alpaca_server.zip 

The example_simulator compile fine and can connect from ccdciel.
I now I look to make the http header available, probably in a TStringList.

**Comment by han-k59** [comment]
Thanks Patrick, Yes I understand from the Ascom developer forum that base64 is hidden in the HTTP header. That is too complicated for me.  I have also noticed an inefficiency in the base64 transmission. This is what I posted at the Ascom developer forum:

----------------------------------------------------------------------------------------------------
There is a factor 2 inefficiency in the base64 communication between AscomRemote and CCDCiel. If I select the “Camera V2 simulator” then type=int32 is selected.  That is not a problem for transmitting the pixel values as string with a comma as separator. However the same happen with the base64 communication!! So the base64 string could be reduced with a factor 2 !!!

Using the standard 800x600 pixel image of the simulator, there are

800 x 600 dimensions ==> 480000 pixels.

The base64 data transmitted by AscomRemote  is 2560000 characters long (tested).

So the efficiency of the transmission is 2560000/480000 is 5.333333 characters/pixel.

Since the efficiency of base64 is 4/3 since 3 bytes can be packed with 4 characters of 6 byte resolution. So 5.333 indicated that 32 bit pixel values are transmitted since (4/3)* 4 bytes is 5.3333 !!

So it would make sense to check if the data is 16 bit or 32 bit before packing it in base64 even if the driver specifies int32. And then transmit it as “int16 & base64”

I know Alpaca is a transmission protocol and maybe the ASCOM driver is to blame but it is better for the standard transmission to use type=int16 since somebody could decide to use the full 32 bit range for int32.

At the moment CCDCiel only knows type=int32 transmission.

Han

**Comment by pchev** [comment]
Han,

The reason to support only int32 is because ASCOM define the imagearray as int32.

I agree that all existing camera are 16bit and this is a wast of resource. But fixing that probably require a whole upgrade of the ASCOM interface.

Also I not see the base64 encoding as a necessity here because the data are not embedded in json but send on a separate http request. In this case a binary transfer of the array is also possible.

**Comment by han-k59** [comment]
Are you saying that the base64 string is not packed in json?  I thought it was be send by a second request Get .... imagearray64: 

_The base64string is obtained by an HTTP GET the endpoint:  
/api/v1/camera/x/imagearraybase64 
where x is the camera device number as used in the original GET to the device’s imagearray 
endpoint._ 

Is your idea of binary transfer already considered by the Alpaca team?

**Comment by pchev** [comment]
Yes there is no json for the base64 data, it is transferred with "Content-Type: image/tiff". This make a binary transfer even it is not a real tiff file.
After ccdciel receive the buffer it decode the base64 data to get the array in binary format.

I see Peter say this is not a finalized way to transfer the data and it's preference is for FITS.
I think this is a good idea because with FITS you no more have to invent a new way to transfer 16bit or 8bit array.

Maybe it is not worth to implement this current base64 method if it is not supported in the future.

**Comment by han-k59** [comment]
Today I tested the transfer (download) and decoding (reading) of  a large image 4656 x3520 pixels.

For a localhost connection

Json int32 strings:

 3 sec transfer
17 sec decoding
20 sec total

Base64 

3 sec transfer 
3 sec decoding 
6 sec total

The data transfer from the Pascal based server was almost instant. Timing was done manually without code.

So the problem lies in the encoding/decoding. So compression is not so important. Encoding and decoding should be simple for speed. Having a Json strings of unequal length separated by commas is slowing down the process significantly. Having a hex string per row would have been better. 

To check if int16 is possible, i did a test and can find the maximum pixel value of the image in 0.05 seconds only.

Han

**Comment by pchev** [comment]
This is interesting, JSON array decoding is really the bottleneck.
In CCDciel I do that using FPC TJSONArray, cu_ascomrest.pas at line 560. 
It is very convenient but maybe this is not the fastest. I can try to replace it by a more simple parser.

**Comment by pchev** [comment]
I add some timing print in CCDciel image array function (when not using base64).
This is with a 4000x3000 image in the simulator, the JSON with the array is 35 MB. The computer is a 15 years old laptop with a Pentium Dual at 2.0 GHz.

The processing of the TJSONArray loop I mention in previous message take only 1.4 second.
Most of the time, 11.5 seconds, is to load the HTTP document from a memory stream to the JSON document. This is at line 370 of cu_ascomrest.pas, FPC function GetJSON.

I will try what it do when parsing the array directly from the memory stream.

**Comment by han-k59** [comment]
Good you have identified the place.

I have now a working "Sky simulator for Alpaca" in Linux running with your server code. A new server problem occurs. Downloading small images of 200x100 pixels via Alpaca link to CCDCiel works, but as soon I try to download images of 300 x 300 pixels or larger, I get the following error in CCdCiel:

22:31:57 camera/0: Error accessing ImageArray: 127.0.0.1:11111 Connection reset by peer

Must be something about the package size or timeout. Something for tomorrow.

Regards, Han

**Comment by han-k59** [comment]
Increasing the timeout doesn't solve the problem. There is unnecessary code in cu_tcpserver:

              if lastError <> 0 then break;
               break;

**Comment by han-k59** [comment]
Removing the second break (twice in the code) solves the problem. I can now transfer large images :)

I have also returned the timeout to the original settings. I don't understand why it is not possible to increase the timeout for:
body:=body+RecvPacket(1); 
but increasing it makes it worse.

So I can now release my simulator for Linux.
https://sourceforge.net/projects/sky-simulator/

I will close this topic. 

Thanks, Han

**Comment by pchev** [comment]
Removing this break create a huge memory leak because the thread never terminate.
It probably just need bit more time to finish to send the data before to close the socket.
I look at that.

**Comment by pchev** [comment]
This must be fixed by eb69eadc1bff20d938365322c4f91ce73f8a9e95 
Can you try with this version.

**Comment by pchev** [comment]
Another fix that prevent huge memory consumption when logging is active: f2e86451cae571e8955c9252983c3b20127324d3

With the previous version it try to show the full json image array in the log, locking the application for a long time.

**Comment by pchev** [comment]
I confirm the sky_simulator for Linux work fine with this two change.

**Comment by han-k59** [comment]
Thanks Patrick, Yes these fixes solve everything. No more disconnect or other problems.  If I put CCDCiel preview in a loop it keeps on working. :)

I fixed some other things in the simulator in the last hour.  

Thanks.

**Comment by han-k59** [comment]
The only thing I was wondering about the short timeout values of 1 ms in cu_tcpserver. Will this work for a poor WiFi connection?

 buf:=RecvString(1);  

body:=body+RecvPacket(1);

**Comment by pchev** [comment]
This work for buf:=RecvString(1); because this only read one header row at a time, all the header are very small and fill in one block so they are already in the input buffer. 
But I can increase the timeout without performance effect because the timeout is normally never used, the normal loop exit is the empty header row:  if trim(buf)='' then break;

This also work for RecvPacket(1) because the JSON data for Alpaca PUT operation is always very small. 
But here it is better if I change it to use the Content-Length header.
CCDciel Alpaca client already use this to read the image data.

This two change are in 32b1b7be4234626a704ca7ca9105799b6d2e2cad
